tools/trap/trap.c

```diff
@@ -288,6 +289,67 @@ static int extfsdump_recv(int dev_fd)
         return 0;
 }
 
+int dump_file_from_fs(int dev_fd, char *filename)
+{
+	int ret;
+	int fd;
+	int i, last_subdir;
+	char buf;
+	FILE *fp;
+	u_int64_t file_size;
+	char char_filesize[8];
+	char local_filename[CONFIG_SMARTFS_MAXNAMLEN + 9];
+
+	printf("Attempting to Dump File %s from Target Device\n", filename);
+
+	ret = write(dev_fd, filename, CONFIG_SMARTFS_MAXNAMLEN);
+	if (ret != CONFIG_SMARTFS_MAXNAMLEN) {
+		printf("Unable to send Dump File name to Target Device\n");
+		return -1;
+	}
+
+	file_size = 0;
+	for (int i = 0; i < 8; i++) {
+		ret = read(dev_fd, &char_filesize[i], sizeof(char_filesize[i]));
+		if (ret != sizeof(char_filesize[i])) {
+			printf("Unable to get file size from Target Device\n");
+			return -1;
+		}
+	}
+
+	file_size = (((u_int64_t)char_filesize[0] << 56) & 0xFF00000000000000U) | (((u_int64_t)char_filesize[1] << 48) & 0x00FF000000000000U) | (((u_int64_t)char_filesize[2] << 40) & 0x0000FF0000000000U) | (((u_int64_t)char_filesize[3] << 32) & 0x000000FF00000000U) | ((char_filesize[4] << 24) & 0x00000000FF000000U) | ((char_filesize[5] << 16) & 0x0000000000FF0000U) | ((char_filesize[6] <<  8) & 0x000000000000FF00U) | (char_filesize[7] & 0x00000000000000FFU);
```

Comment:
Instead of this tricky code, you can use uint64_t buf when you write & read.
write file size : write(fd, &buf, sizeof(uint64_t));
get file size : read(fd, &buf, sizeof(uint64_t));
anyway, is file_size written in header of file?

Reply:
Because you write file name and readback, then calculate file size.
That's why asking you is there information regarding file size at the start of file.
I can't understand why this code as added because we can use stat to get size of file easily.

Reply:
We need file_size so that we know how many characters of file data we are supposed to receive from the crashed device.
We get this file size from the device itself after we pass the name of the file that we want.
The device side performs a stat for us to get the file size.
However, the device sends back one character at a time using up_lowputc, but we need 8 characters (64 bytes) in total for file_size.
Hence, we receive one character at a time and combine them using this logic into a 64 bit file_size.

We cannot perform a stat to get the userfs file size here (in trap.c) because this file is run on the Linux host machine